### DISCUSSION TITLE
Next.js zustand context with .getState()

### DISCUSSION THREAD

**Discussion by ootkin** [pmndrs/zustand#2315] [Bug report]
Hi folks,

I'm using zustand in order to initialize the state in a Next.js app with RSC.

To do that, I followed the guide in your docs, in particular: https://docs.pmnd.rs/zustand/previous-versions/zustand-v3-create-context#migration

However, with `useStore(store, selector)`is not possible to use the `.getState()` from the hook, and so it is not possibile to set a state outsite a react component:

```ts
function getUserOutsideReactComponent() {
    const user = useUserContext(state => state)
}
``` 

In this case `useUserContext(state => state)` throw the common error: "Error: Invalid hook call. Hooks can only be called inside of the body of a function component" because it is outside of a RC, and it does not have `.getState()`.

Is there a way to retrieve the state outsite of a RC if we are using React context?

Thanks

**Comment by dai-shi** [comment]
> I'm using zustand in order to initialize the state in a Next.js app with RSC.

Using the Zustand state on the server means that it's just a global variable across sessions. It's generally not recommended.

To initialize Zustand state on the client from the server, we should use a specific pattern. #2298 is a WIP PR for it.

**Reply by ootkin** [reply]
Sorry if I didn't explain myself correctly.

The problem I'm trying to solve is not the use of Zustand state on the server side, but to use the `.getState()` method aiming to retrieve the state outside of a React component.

From this example:
```ts
function useBearContext<T>(selector: (state: BearState) => T): T {
  const store = useContext(BearContext)
  if (!store) throw new Error('Missing BearContext.Provider in the tree')
  return useStore(store, selector)
}
```
The `useBearContext` hook returns the `useStore` hook, which does not have the `.getState()` method, and therefore it's not possible to receive the state in this way:
```ts
const state = useBearContext.getState()
```

**Reply by dai-shi** [reply]
Oh, sorry, I totally misunderstood.

Here's something I would do:

```js
function useBearStore() {
  const store = useContext(BearContext)
  if (!store) throw new Error('Missing BearContext.Provider in the tree')
  return store
}
```

**Reply by ootkin** [reply]
Mmh, but in this way I cannot pass a selector and i cannot destructure the result of the hook...

**Reply by dai-shi** [reply]
Do you mean something like this?
```js
function useBearStore(selector) {
  const store = useContext(BearContext)
  if (!store) throw new Error('Missing BearContext.Provider in the tree')
  return selector(store.getState())
}
```

**Reply by ootkin** [reply]
Yes, but in this way i cannot call `useBearStore().getState()` or `useBearStore().setState({})` if I want to use the store outside of a react component.

**Reply by dai-shi** [reply]
You can't access context anyway outside react, because context only exists within a react tree.

**Reply by ootkin** [reply]
You are right, so there is no way to call setState and getState on a nextjs app using context, is it correct?

**Reply by dbritto-dev** [reply]
@ootkin yes, on nextjs since you need to use context you can't use it outside of react components/hooks

**Reply by ootkin** [reply]
Ok, it makes sense.

I've got another question, looking at the documentation about setup with next: https://docs.pmnd.rs/zustand/guides/nextjs.

The code is working great on my project, the problem is that we need to write a lot of boilerplate code in order to setup the zustand context properly. Now i have like 10 different zustand stores, and for each store we need to write the same boilerplate.

Is there a way create a wrapper that creates context and providers in a more clean and elegant way?

**Reply by dbritto-dev** [reply]
@ootkin sure, zustand guide was created for beginners but there's some patterns that you can use to avoid the boilerplate. Do you have an example about what are you trying to achieve?

**Reply by ootkin** [reply]
The problem is that I need to get the state of a store inside an action of another store and i cannot use hooks,

is there a workaround?

**Reply by dbritto-dev** [reply]
@ootkin when there's a dep between two states the best option is merge them into one single state and use the slice pattern

**Reply by ootkin** [reply]
@dbritto-dev Ok, thank you.

As soon as I have some time, I will create a repository with examples addressing the boilerplate questions. Would it be better to start a new discussion and close this one?

**Reply by dbritto-dev** [reply]
@ootkin just start a new one

**Reply by TimurKr** [reply]
@ootkin, I was struggling with similar issues and figured a way of creating a singular store with slices patterns and provide it using context. I created a library to get rid of as much of the boilerplate as possible. Check it out, it might give you some answers: https://github.com/TimurKr/zimmer-context

**Reply by dbritto-dev** [reply]
https://github.com/pmndrs/zustand/discussions/2326#discussioncomment-8519935
